python_modules/dagster/dagster/_core/definitions/asset_checks.py

```diff
@@ -0,0 +1,119 @@
+from typing import Any, Dict, Iterable, Iterator, Mapping, Set
+
+from dagster import _check as check
+from dagster._annotations import experimental, public
+from dagster._core.definitions.asset_check_spec import AssetCheckSpec
+from dagster._core.definitions.events import AssetKey
+from dagster._core.definitions.node_definition import NodeDefinition
+from dagster._core.definitions.resource_definition import ResourceDefinition
+from dagster._core.definitions.resource_requirement import (
+    ResourceAddable,
+    ResourceRequirement,
+    merge_resource_defs,
+)
+
+
+@experimental
+class AssetChecksDefinition(ResourceAddable):
+    """Defines a set of checks that are produced by the same op or op graph.
+
+    AssetChecksDefinition are typically not instantiated directly, but rather produced using a
+    decorator like :py:func:`@asset_check <asset>`.
+    """
+
+    _node_def: NodeDefinition
+    _resource_defs: Mapping[str, ResourceDefinition]
+    _specs_by_output_name: Mapping[str, AssetCheckSpec]
+
+    def __init__(
+        self,
+        node_def: NodeDefinition,
+        resource_defs: Mapping[str, ResourceDefinition],
+        asset_keys_by_input_name: Mapping[str, AssetKey],
+        specs_by_output_name: Mapping[str, AssetCheckSpec],
```

Comment:
I'd prefer a having a separate data structure to handle the input/output mappings in terms of a public API. That's also where I think `AssetsDefinition` to go. 

```
def __init__(
  self, 
  node_def,
  resource_defs,
  asset_keys: Sequence[AssetKey], # note: can we make this asset specs as well? probably not..
  asset_check_specs: Sequence[AssetCheckSpec], 
  input_output_mapping, # new object the encodes the input/output mappings
)
```

This will cause a little more bookkeeping in the code that creates the definitions object from the decorator, I think it provides optional value for a cleaner API in post i/o manager world. The input/output stuff with the above parameters ends up feeling like a "layer" on top of the core asset_keys, and asset_check_specs. This provides option value for the underlying framework to handle output/input shenanigans on the user's behalf, and making it optional.

At some point I am very interested in a refactor to do the same to `AssetsDefinition` to eliminate the nest of mapping dictionaries there and make it more rationale. I'd rather get ahead of it here.

Curious re: @alangenfeld's take on this.

Reply:
interested to hear how you landed on `specs_by_output_name` to see if I am missing something but otherwise inclined to agree with Nick and would prefer `specs: Sequence`

Reply:
@alangenfeld take a look at https://github.com/dagster-io/dagster/pull/15930/files#r1298828398

Reply:
> This provides option value for the underlying framework to handle output/input shenanigans on the user's behalf,

@schrockn I'm having a little trouble picturing this. Do you have an example in mind of what this could look like?

Reply:
One way of looking at an `AssetChecksDefinition` is as an adapter between the `NodeDefinition` data model and the asset check data model. In the limits of my current imagination, a world where inputs and outputs aren't relevant is a world where either:
- `NodeDefinition`s directly understand asset checks.
- Inputs and outputs aren't required to define edges between node definitions.

Reply:
As a separate point, the approach here already avoids the proliferation of dictionaries that we have on `AssetsDefinition`, because everything goes inside the `AssetCheckSpec` instead of having a separate dictionary per piece of check-level metadata.

Reply:
```
class InputOutputMappings:
   output_names_to_asset_keys: Mapping[str, AssetKey]
   input_names_to_check_names: Mapping[str, str]
```

This would allow the framework to synthesize input and output names on the users behalf if they do not provide this mapping object

Reply:
If I understand correctly, the suggestion is something like this:

```python
from typing import Mapping, Sequence
from dagster import AssetKey, AssetCheckSpec


class AssetCheckHandle:
    asset_key: AssetKey
    check_name: str


class AssetChecksDefinition:
    def __init__(
        self,
        specs: Sequence[AssetCheckSpec],
        asset_check_handles_by_output_name: Mapping[str, AssetCheckHandle],
    ):
        self._specs_by_handle = {spec.handle: spec for spec in specs}
        self._specs_by_output_name = {
            output_name: self._specs_by_handle[check_handle]
            for output_name, check_handle in asset_check_handles_by_output_name.items()
        }
```

Reply:
I'm proposing a single object that takes responsibility of mapping input/output space to asset space (key and check)

Reply:
Yeah Sandy that is the spirit for the output side of things.

Reply:
Updated the PR to include `AssetChecksDefinitionInputOutputProps`. I avoided the word "mapping" for fear of confusion with `InputMapping` and `OutputMapping`.